lib/clerk/authenticate_request.rb

```diff
@@ -0,0 +1,245 @@
+                return signed_out if !token
+
+                if token["iat"] < auth_context.client_uat
+                    return handle_handshake_maybe_status(env, reason: AuthErrorReason::SESSION_TOKEN_OUTDATED);
+                end
+
+                return signed_in(env, token, auth_context.session_token_in_cookie)
+            rescue JWT::ExpiredSignature
+                handshake(env, reason: TokenVerificationErrorReason::TOKEN_EXPIRED)
+            rescue JWT::InvalidIatError
+                handshake(env, reason: TokenVerificationErrorReason::TOKEN_NOT_ACTIVE_YET)
+            end
+
+            signed_out
+        end
+
+        def resolve_handshake(env)
+            headers = { 
+                "Access-Control-Allow-Origin" => "null",
+                "Access-Control-Allow-Credentials" => "true"
+            }
+            
+            session_token = ''
+
+            handshake_payload = verify_token(auth_context.handshake_token)
+            return signed_out(enforce_auth: true, reason: TokenVerificationErrorReason::JWK_FAILED_TO_RESOLVE) if !handshake_payload
+
+            cookies_to_set = handshake_payload[HANDSHAKE_COOKIE_DIRECTIVES_KEY] || []
+            cookies_to_set.each do |cookie|
+                headers[COOKIE_HEADER] ||= []
+                headers[COOKIE_HEADER] << cookie
+
+                if cookie.start_with?("#{SESSION_COOKIE}=")
+                    session_token = cookie.split(';')[0].split('=')[1]
+                end
+            end
+
+            if auth_context.development_instance?
+                redirect_url = auth_context.clerk_url.dup
+                remove_from_query_string(redirect_url, HANDSHAKE_COOKIE)
+                remove_from_query_string(redirect_url, HANDSHAKE_HELP_QUERY_PARAM)
+
+                headers[LOCATION_HEADER] = redirect_url.to_s
+            end
+            
+            if !session_token
+                return signed_out(reason: AuthErrorReason::SESSION_TOKEN_MISSING, headers: headers)
+            end
+
+            begin
+                claims = verify_token(session_token)
+                return signed_in(env, claims, session_token) if claims
+            rescue JWT::ExpiredSignature, JWT::InvalidIatError => e
+                if auth_context.development_instance?
+                    # TODO: log possible Clock skew detected
+                
+                    # Retry with a generous clock skew allowance (1 day)
+                    claims = verify_token(session_token, timeout: 86_400)
+                    return signed_in(env, claims, session_token) if claims
+                end
+                
+                # Raise error if handshake resolution fails in production
+                raise e
+            end
+        end
+
+        def handle_handshake_maybe_status(env, **opts)
+            return signed_out if !eligible_for_handshake?
+            handshake(env, **opts)
+        end
+
+        # A outcome
```

Comment:
:question: What is this comment about? Perhaps we can abstract the possible outcomes in classes and make sure we return/invoke the correct outcome?

Would that convey the notion of an outcome better in code?

Reply:
I saw after I finished the review that this code has been moved over, along with the comments.

Not suggesting you do anything at this point, unless you're up for a refactoring trip. :smile: :champagne:

Reply:
i will keep it as is for now and we should tackle it in the future if there is a need to refactor it :)

Reply:
Yessir!